blockchain/accept.go

```diff
@@ -90,3 +92,84 @@ func (b *BlockChain) maybeAcceptBlock(block *btcutil.Block, flags BehaviorFlags)
+// them, so those fields are not included here.  This provides support for full
+// headers-first semantics.
+//
+// The flag for check header sanity allows the additional header sanity checks
+// to be skipped which is useful for the full block processing path which checks
+// the sanity of the entire block, including the header, before attempting to
+// accept its header in order to quickly eliminate blocks that are obviously
+// incorrect.
+//
+// In the case the block header is already known, the associated block node is
+// examined to determine if the block is already known to be invalid, in which
+// case an appropriate error will be returned.  Otherwise, the block node is
+// returned.
+//
+// This function MUST be called with the chain lock held (for writes).
+func (b *BlockChain) maybeAcceptBlockHeader(header *wire.BlockHeader, checkHeaderSanity bool) (*blockNode, error) {
+	// Avoid validating the header again if its validation status is already
+	// known.  Invalid headers are never added to the block index, so if there
+	// is an entry for the block hash, the header itself is known to be valid.
+	// However, it might have since been marked invalid either due to the
+	// associated block, or an ancestor, later failing validation.
+	hash := header.BlockHash()
+	if node := b.index.LookupNode(&hash); node != nil {
+		if err := b.checkKnownInvalidBlock(node); err != nil {
+			return nil, err
+		}
+
+		return node, nil
+	}
+
+	// Perform context-free sanity checks on the block header.
+	if checkHeaderSanity {
+		err := checkBlockHeaderSanity(header, b.chainParams.PowLimit, b.timeSource, BFNone)
+		if err != nil {
+			return nil, err
+		}
+	}
+	// Orphan headers are not allowed and this function should never be called
+	// with the genesis block.
+	prevHash := &header.PrevBlock
+	prevNode := b.index.LookupNode(prevHash)
+	if prevNode == nil {
+		str := fmt.Sprintf("previous block %s is not known", prevHash)
+		return nil, ruleError(ErrMissingParent, str)
+	}
+
+	// There is no need to validate the header if an ancestor is already known
+	// to be invalid.
+	prevNodeStatus := b.index.NodeStatus(prevNode)
+	if prevNodeStatus.KnownInvalid() {
+		str := fmt.Sprintf("previous block %s is known to be invalid", prevHash)
+		return nil, ruleError(ErrInvalidAncestorBlock, str)
+	}
+
+	// The block must pass all of the validation rules which depend on the
+	// position of the block within the block chain.
+	err := b.checkBlockHeaderContext(header, prevNode, BFNone)
+	if err != nil {
+		return nil, err
+	}
+
+	// Create a new block node for the block and add it to the block index.
+	//
+	// Note that the additional information for the actual votes, tickets, and
+	// revocations in the block can't be populated until the full block data is
+	// known since that information is not available in the header.
+	newNode := newBlockNode(header, prevNode)
+	newNode.status = statusNone
+	b.index.AddNode(newNode)
+	err = b.index.flushToDB()
```

Comment:
So looking into this more, I'd prefer having the flush happening in `ProcessBlockHeader` since we'd eventually want to be able to re-use this function in `ProcessBlock` and `ProcessBlock` takes care of the flush itself.

Reply:
okay that makes sense, since we would be using maybeAcceptBlockHeader in ProcessBlock when we would receive a block whose header haven't been validated.

Reply:
But do we need header validation, if we have the complete block data to us and we are using maybeAcceptBlock function to validate block anyways.

Reply:
Anyways I will add the dbFlush to ProcessBlockHeader function